server/src/main/java/org/elasticsearch/search/aggregations/metrics/CardinalityAggregatorFactory.java

```diff
@@ -48,7 +49,28 @@
     }
 
     public static void registerAggregators(ValuesSourceRegistry.Builder builder) {
-        builder.register(CardinalityAggregationBuilder.REGISTRY_KEY, CoreValuesSourceType.ALL_CORE, CardinalityAggregator::new, true);
+        builder.register(CardinalityAggregationBuilder.REGISTRY_KEY,
+            CoreValuesSourceType.ALL_CORE,
+            (name, valuesSourceConfig, precision, context, parent, metadata) -> {
+                // super hacky but it shows the point of the approach
+                if (valuesSourceConfig.hasValues()) {
+                    ValuesSource valuesSource = valuesSourceConfig.getValuesSource();
+                    if (valuesSource instanceof ValuesSource.Bytes.WithOrdinals) {
+                        ValuesSource.Bytes.WithOrdinals source = (ValuesSource.Bytes.WithOrdinals) valuesSource;
+                        // is this call expensive? I hope not
+                        final long maxOrd = source.globalMaxOrd(context.searcher());
```

Comment:
My understanding is that this call is generally cheap if the global ords are built (e.g. getting the max ord is a cheap operation).  But this will trigger rebuilding ords if the segments have changed and can then be very expensive :)

Does Lucene have a cheap way to get the total number of terms for a field?  Even though that doesn't factor in deleted docs or any filtering, perhaps that'd be sufficient heuristic?

Reply:
I don't think we want to trigger rebuilding rods if we are not going to use them

Reply:
Yes we had the same problem in the `terms` aggregation but users can set the mode to `map` to avoid loading the global ordinals. 

> Does Lucene have a cheap way to get the total number of terms for a field? Even though that doesn't factor in deleted docs or any filtering, perhaps that'd be sufficient heuristic?

Lucene provides the number of different terms in SortedSetDocValues but that's a per-segment view. The global view is provided by global ordinals but they can be costly to compute. The cardinality in each segment could be enough for the heuristic though, you can take the max or the sum and avoid loading the global ordinal until you have some hints that it might save some significant memory ?

Reply:
Yes, probably we need some heuristics considering as well the ratio of number of docs / number of different terms as it will give us an idea if there is many repeated values?